genienlp/ned/bootleg.py

```diff
@@ -135,116 +137,110 @@ def extract_mentions(self, input_path):
     def disambiguate_mentions(self, config_args):
         run_model(self.args.bootleg_dump_mode, config_args)
 
-    def post_process_bootleg_types(self, title):
-        types = None
-        title = title.lower()
-        for pair in self.wiki2normalized_type:
-            if fnmatch.fnmatch(title, pair[0]):
-                types = pair[1]
-                break
-
-        typeqids = None
-        if types is not None:
-            if isinstance(types, str):
-                typeqids = [self.type_vocab_to_typeqid[types]]
-            elif isinstance(types, (list, tuple)):
-                typeqids = [self.type_vocab_to_typeqid[type_] for type_ in types]
-
-        return typeqids
-
-    def collect_features_per_line(self, line, threshold):
+    def collect_features_per_alias(self, alias, all_probs, all_qids):
+        type_ids = []
+        type_probs = []
+        qids = []
+
+        # sort candidates based on bootleg's confidence scores (i.e. probabilities)
+        # this used to be in bootleg code but was removed in recent version
+        packed_list = zip(all_probs, all_qids)
+        packed_list_sorted = sorted(packed_list, key=lambda item: item[0], reverse=True)
+        all_probs, all_qids = list(zip(*packed_list_sorted))
+
+        # filter qids with probability lower than a threshold
+        idx = reverse_bisect_left(all_probs, self.args.bootleg_prob_threshold)
+        all_qids = all_qids[:idx]
+        all_probs = all_probs[:idx]
+
+        if len(all_qids) > self.args.max_qids_per_entity:
+            all_qids = all_qids[: self.args.max_qids_per_entity]
+            all_probs = all_probs[: self.args.max_qids_per_entity]
+
+        if not is_banned(alias):
+            for qid, prob in zip(all_qids, all_probs):
+                # to map qids to unique ids we just need to remove the Q character as qids are distinct
+                qids.append(int(qid[1:]))
+
+                # get all types for a qid
+                all_types = []
+                if 'books' in self.args.ned_domains:
+                    if alias in ['houghton mifflin', 'ciudad de buenos aires']:
+                        all_types = ['award']
+                    elif alias in ['blanche', 'biblioteca breve', '3rd edition', '4th edition']:
+                        all_types = ['version, edition, or translation']
```

Comment:
This is not the place for this stuff.

Also, to have this stuff at all in Bootleg is meaningless: you're just raising your accuracy on the dev set by looking at oracle results and cleaning up by hand, it's not scientific.

Reply:
I'm reaching the conclusion that for certain entities we need to map them directly to the correct type.
It happens in 2 cases:
1) bootleg partly detected the entity and thus got the Qid wrong. 
2) bootleg alias2qids has incorrect entries where the book’s writer Qid is associated with the book name instead of the book itself, thus mapping to “person” type instead of “book”.
In none of those cases changing type mapping will fix the problem. Also, this mapping is only done for training set, not dev/ test set.
I added this as if-else in the code to see how experiments turn out. I'm eventually gonna move it to a json file similar to type mapping.

Reply:
Yeah don't merge this hacks to the code. Add an override file if you need to, but I still don't think you need to.